### PR TITLE
global: redesign of the Obelix package

### PR DESCRIPTION
**Pull request by ddaze**
- Separates Obelix into a Python package.
- Creates setup.py file for the package.
- Moves from Redis-only storage to generic Storage class.
- Implements a subclass of Storage that uses Redis.
- Decouples the logging from using Redis directly to a generic Signal class.
- Moves helper functions to utils.
- Rewrites tests to match the new code structure.

### CODE REVIEW

**Comment by ddaze** [comment]
Comments and ideas to the design and implementation are more than welcome.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+*.pyc
+build/
+dist/
+obelix_client.egg-info/
+.cache/
+.coverage
+.eggs/
+pytest_cov-1.8.1-py2.7.egg/
```
Comment (jirikuncar) [inline .gitignore]: what about `*.egg`?

File: obelix_client/__init__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# This file is part of Invenio.
```
Comment (jirikuncar) [inline obelix_client/__init__.py]:
```
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
```

File: obelix_client/api.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                        'jrec': jrec,
+                        'rg': rg,
+                        'rm': rm,
+                        'cc': cc,
+                        'hit_number_local': jrec + hit_number_local,
+                        'hit_number_global': jrec + hit_number_global,
+                        'recommendations': recommendations,
+                        'recid_in_recommendations': recid in recommendations,
+                        'type': type}
+                Obelix.signal_statistics_page_view.send(self, data=data)
```
Comment (jirikuncar) [inline obelix_client/api.py]: You should be able to use `self.signal_statistics_page_view.send` (easier subclassing)

## 

PS: Please have a look to other occurrences too.

File: obelix_client/queue.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (jirikuncar) [inline obelix_client/queue.py]: missing license header

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Invenio is a framework for digital libraries and data repositories.
```
Comment (jirikuncar) [inline setup.py]: update with Obelix Client description ...

File: obelix_client/__init__.py (hunk 2 above)
Comment (ddaze) [inline obelix_client/__init__.py]: change it to: # This file is part of Obelix.   ?
Would it be enough to create a License file and remove the header from the other files, or just keep it in **init**.py or api.py ?

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Obelix.
+# Copyright (C) 2013, 2014, 2015 CERN.
```
Comment (egabancho) [inline .travis.yml]: Just the current year if new file

File: obelix_client/queue.py (hunk 4 above)
Comment (ddaze) [inline obelix_client/queue.py]: Done

File: .travis.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#
+# Invenio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
```
Comment (jirikuncar) [inline .travis.yml:19]: please add `sudo: false` to use new infrastructure

File: obelix_client/api.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Obelix; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Obelix-Client Search Engine """
+
+import time
+import logging
+import re
+import obelix_client.utils as utils
```
Comment (jirikuncar) [inline obelix_client/api.py]: `import .utils as utils` / `from . import utils`

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        import pytest
+        import _pytest.config
+        pm = _pytest.config.get_plugin_manager()
+        pm.consider_setuptools_entrypoints()
+        errno = pytest.main(self.pytest_args)
+        sys.exit(errno)
+
+setup(
+    name='obelix-client',
+    version=0.1,
```
Comment (jirikuncar) [inline setup.py]: please move it to `obelix_client/version.py:0.1.0.devYYYYMMDD`

**Comment by egabancho** [comment]
> - Travis CI build

You are missing a dot at the end of the line here :wink:

File: obelix_client/api.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+CONFIG = {'recommendations_impact': 0.5,
+          'score_lower_limit': 0.2,
+          'score_upper_limit': 10,
+          'score_min_limit': 10,
+          'score_min_multiply': 4,
+          'score_one_result': 1,
+          'method_switch_limit': 20,
+          'user_identifier': 'uid',
+          }
```
Comment (ddaze) [inline obelix_client/api.py]: pylint is complaining with:
C: 37, 0: Wrong continued indentation.
          }
         |^ (bad-continuation)

but when I change it, pep8 is complaining with:
obelix_client/api.py:37:10: E124 closing bracket does not match visual indentation

@jirikuncar

Comment (jirikuncar) [inline obelix_client/api.py]: Usual style is:

``` python
CONFIG = {
    'foo': 1,
    'bar': 2,
}
```

Comment (ddaze) [inline obelix_client/api.py]: Thank you!

File: obelix_client/api.py [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+def get_logger():
+    """Get a Logger."""
+    return logging.getLogger('obelix_client')
+
+
+class Obelix(object):
+
+    """Obelix-Client."""
+
+    def __init__(self, cache_storage, recommendation_storage, queue_storage,
```
Comment (ddaze) [inline obelix_client/api.py]: @jirikuncar pylint is warning about to many arguments, does it make sense to do it different, if yes, how?
Thank you!

R: 50, 4: Too many arguments (6/5) (too-many-arguments)

**Comment by ddaze** [comment]
Ping @jirikuncar

**Pull request merged**